vendor/assets/javascripts/mountable_file_server.js

```diff
@@ -1,61 +1,44 @@
 (function() {
-  var lastUploadId = 0;
-  var finishedUploads = [];
-
-  var dispatchEvent = function($element, name, payload) {
-    var event = document.createEvent('CustomEvent');
-
-    event.initCustomEvent(name, true, false, payload);
-    $element.dispatchEvent(event);
-  };
-
-  var uploadFile = function($element, file, uploadId) {
-    var url = $element.getAttribute('data-endpoint');
-    var type = $element.getAttribute('data-type');
-    var xhr = new XMLHttpRequest();
-    var formData = new FormData();
-
-    formData.append('file', file);
-    formData.append('type', type);
-
-    xhr.open('POST', url, true);
-
-    xhr.onreadystatechange = function() {
-      if (xhr.readyState === 4 && xhr.status === 200) {
-        finishedUploads.push(uploadId);
-
-        response = JSON.parse(xhr.responseText);
-
-        dispatchEvent($element, 'upload:success', {
-          uploadId: uploadId,
-          uid: response.uid,
-          wasLastUpload: lastUploadId == finishedUploads.length
-        });
-      }
-    }
-
-    xhr.upload.addEventListener('progress', function(progressEvent) {
-      if (progressEvent.lengthComputable) {
-        dispatchEvent($element, 'upload:progress', {
-          uploadId: uploadId,
-          progress: progressEvent
+  var MountableFileServerUploader = function(options) {
+    return {
+      url: options.url,
+      uploadFile: function(options) {
+
+        var file = options.file;
+        var type = options.type;
+        var onStart = options.onStart || function() {};
+        var onProgress = options.onProgress || function() {};
+        var onSuccess = options.onSuccess || function() {};
```

Comment:
Just another thought out of the blue: Is there a use case for a callback `onError`? If yes, we could add it.